protobuf/lib/src/protobuf/generated_message.dart

```diff
@@ -215,24 +222,65 @@ abstract class GeneratedMessage {
   /// represented as their integer value.
   ///
   /// For the proto3 JSON format use: [toProto3Json].
-  String writeToJson() => jsonEncode(writeToJsonMap());
+  String writeToJson() {
+    final buf = StringBuffer();
+    final stringSink = jsonStringWriter(buf);
+    writeToJsonSink(stringSink);
+    return buf.toString();
+  }
+
+  void writeToJsonSink(JsonSink jsonSink) {
```

Comment:
Do we need to expose this?

Not that I expect the interface to change anytime soon, but exposing api from another package is always extra liability, and as such we should prefer to keep usage internal.

Reply:
Good point re: exposing API from another library. No one is requesting this at the moment, and users start using the sink API it will be difficult to migrate to another library later, or maybe drop jsontool entirely.

OTOH if needed we can always add it later without breakage (other than messages with field name `write_to_json_sink`).

I'll remove this method.

Reply:
I removed `writeToJsonSink`, but we can't remove `toProto3JsonSink` because it's called from `AnyMixin`s `toProtoJsonHelper`, which now takes a sink argument as well to allow reuse from entry points `toProto3JsonString` and `toProto3JsonObject`. I think the best we can do is hiding `toProto3JsonSink` in documentation.

Reply:
It's definitely a liability to expose a third-party package's API in your API. If they change the API, it affects your API too, which means you might need a tight version constraint.

If the method can be hidden in documentation, it's not clear that users ever need to call it. Could it be made private, and only accessible by calling a special static helper function, like:
```dart
 .... 
 dynamic _toProto3JsonObject(...) ...
}

dynamic toProto3JsonObject(YourClass o) => o._toProto3JsonObject();
```
and then don't export that function in the public API (`hide` it), just rely on it in `AnyMixin`.